From b71ac42c6e18e02785197a92c0508763a340d8e8 Mon Sep 17 00:00:00 2001
From: Dmitry Kasatkin <d.kasatkin@samsung.com>
Date: Fri, 17 Jan 2014 14:47:57 +0200
Subject: [PATCH 38/45] evm: do not use inode generation for special files

There is no way to read inode generation number for special files.
It makes it impossible to perform filesystem labeling with digital
or HMAC signatures. Disable it for time being.

Signed-off-by: Dmitry Kasatkin <d.kasatkin@samsung.com>
---
 security/integrity/evm/evm_crypto.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/security/integrity/evm/evm_crypto.c b/security/integrity/evm/evm_crypto.c
index 7cd3724..4e1482c 100644
--- a/security/integrity/evm/evm_crypto.c
+++ b/security/integrity/evm/evm_crypto.c
@@ -123,7 +123,11 @@ static void hmac_add_misc(struct shash_desc *desc, struct inode *inode,
 
 	memset(&hmac_misc, 0, sizeof(hmac_misc));
 	hmac_misc.ino = inode->i_ino;
-	hmac_misc.generation = inode->i_generation;
+	/* inode generation can be read from user space only
+	 * for files and directories
+	 */
+	if (S_ISREG(inode->i_mode) || S_ISDIR(inode->i_mode))
+		hmac_misc.generation = inode->i_generation;
 	hmac_misc.uid = from_kuid(&init_user_ns, inode->i_uid);
 	hmac_misc.gid = from_kgid(&init_user_ns, inode->i_gid);
 	hmac_misc.mode = inode->i_mode;
-- 
2.1.4

